src/main/java/uk/gov/companieshouse/api/accounts/interceptor/CompanyAccountInterceptor.java

```diff
@@ -0,0 +1,101 @@
+ * transaction, checks the account ID is the same as the one that is referenced in the relevant
+ * Transaction. This interceptor will fail if the Transaction interceptor has not run prior to its
+ * own execution.
+ */
+@Component
+public class CompanyAccountInterceptor extends HandlerInterceptorAdapter {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger("company-accounts.api.ch.gov.uk");
+    @Autowired
+    private CompanyAccountService companyAccountService;
+
+    /**
+     * This method extracts the 'company_account' parameter passed via the URI then validates it.
+     * Validation is carried out via a database lookup for a CompanyAccountEntity Object then
+     * matching the retrieved Entities 'self' link to the session stored Transactions
+     * 'company_account' link. Providing this validation passes it assigns the CompanyAccountEntity
+     * to the session.
+     *
+     * @param request - current HTTP request
+     * @param response - current HTTP response
+     * @param handler - chosen handler to execute, for type and/or instance evaluation
+     * @return true if the execution chain should proceed with the next interceptor or the handler
+     * itself.
+     */
+    @Override
+    @SuppressWarnings("unchecked")
+    public boolean preHandle(HttpServletRequest request, HttpServletResponse response,
+            Object handler) {
+        HttpSession session = request.getSession();
+        Transaction transaction = (Transaction) session
+                .getAttribute(AttributeName.TRANSACTION.getValue());
+        if (transaction == null) {
+            LOGGER.error(
+                    "CompanyAccountInterceptor failed on preHandle: Failed to retrieve a transaction from the session.");
+            return false;
+        }
+
+        Map<String, String> pathVariables = (Map) request
+                .getAttribute(HandlerMapping.URI_TEMPLATE_VARIABLES_ATTRIBUTE);
+        String companyAccountId = pathVariables.get("companyAccountId");
+        CompanyAccountEntity companyAccountEntity = companyAccountService
+                .findById(companyAccountId);
+        if (companyAccountEntity == null) {
+            LOGGER.error(
+                    "CompanyAccountInterceptor failed on preHandle: Failed to retrieve a CompanyAccount for "
+                            + transaction
+                            .getId() + ".");
+            return false;
+        }
+
+        String accountsSelf = companyAccountEntity.getData().getLinks()
+                .get(LinkType.SELF.getLink());
+        Map<String, Resources> resourcesList = transaction.getResources();
+        if (!isLinkInResourceMap(resourcesList, accountsSelf)) {
+            LOGGER.error(
+                    "CompanyAccountInterceptor failed on preHandle: Failed to find the CompanyAccount self link in the Transactions links "
+                            + transaction
+                            .getId() + ".");
+            return false;
```

Comment:
each return false case should set the correct status code to report back to the end user, I'd expect this case to be a 404 as the company accounts resource is not found for this transaction

Brain dump:
thought process for the 404 vs some other response code is that they get the same response code if the company account resource doesn't exist at all and if it doesn't exist on the correct transaction meaning that if this gets called with a company account id that exists on a different transaction it doesn't give that fact away to the end user, they just know that it doesn't exist rather than them knowing that it does exist but for a different transaction